### DISCUSSION TITLE
A way to manually add new cars and tracks

### DISCUSSION THREAD

**Discussion by Evansblue787** [abesto/ams2-career#74] [General]
Hi, I love this app and have been using it now for over a month. Some of the content is a little dated so I was wondering if it was possible to add new cars and tracks myself? 

Also going further it would be really awesome if I could even rewrite the cars and tracks for a different game? Say if I wanted to make it relevant to rfactor 2?

Thank you

**Comment by abesto** [comment]
Hi! Always happy to hear you're having fun with it. https://github.com/abesto/ams2-career/issues/67#issuecomment-1374121325 outlines what it'd take to update the "official" car / track list.

If you have some development experience (that's a big if, but I'll document it anyway): it should be fairly easy to make changes for yourself (locally) that you want, either by
* directly editing the `.csv` files in `src/app/data`, or
* making a copy of the Google Sheet (see linked comment above), and using `yarn run update-data` after changing the hard-coded URL in the code

and then running the app locally.

---

To make this fully user-customizable would be tricky: it'd either need a way to resolve conflicts with the built-in data (what if the "official" set later gets the track / car you added?), or the way data is handled today would need to be completely reworked. Not to mention, you'd want to make extra super duper sure you never lose the work put into the track / car set.

After ~10 minutes of thinking, my best idea for how to do this well would look something like this:
* Extend the save data to optionally contain car / track data
* By default the built-in stuff is updated automatically with new versions of the app. You can opt out of this, at which point the current built-in entries get copied into the save data, and you're the sole responsible owner of the car / track data in your save, which means you can edit the ex-"built-in" ones, so you can for example replace the data with rFactor 2 cars / tracks. If you do this, you'll never again get "official" dataset updates on that save.
  * Side note: extending the app to support multiple built-in datasets (like, one for AC, one for rFactor 2, etc.) is not technically hard. The hard part is filling in the data.
* Alternatively, you can stay with the built-in dataset, and add entries via either some simple JSON/CSV text editor (easy to implement, hard to use) or a custom UI (hard to implement, easy to use). This solves adding small pieces of content you miss.
  * This would ideally also come with a button to merge custom tracks / cars into (new) built-in ones for when the built-in dataset catches up, to keep your history sane and to prevent now-duplicated cars / tracks from showing up twice as often.
* There needs to be a way to dump your custom car / track entries. This is trivial to implement. This provides both 1. an easy way to back up and share custom datasets, and 2. send suggestions for data to include in the built-in dataset.

I severely dislike the split this introduces between formats of data storage (Google Sheet vs in-app / JSON), but it's the best I can come up with without introducing a backend component (at which point we could use either Google Sheets of something else as a backing store). This is something I very intentionally avoided, to keep the whole app simple. It means you can keep using it "forever", because there's no server component that needs me to maintain it.

---

Further side note: one reason the data entry is hard is that there's a full matrix of ALL CAR CLASSES x ALL TRACKS VARIATIONS, with a checkbox in each cell. Some ideas around that:
* We might introduce some very high-level categories for both cars and tracks, and define relationships only between those.
* For user-added content, we'd probably want an option to say "just add this car / track, let it race anywhere / let anything race there". This reduces the friction of adding the car / track, and presumably the user can sanity check the combo when it's generated.
  * Follow-up to that: if a combo is raced, maybe mark it as good. Maybe add a button "this combo is insane" to mark it as bad.

---

## OK so will I work on some / any of this?

I'd want a fairly strong commitment from someone to become the steward of the built-in dataset before I invest time in implementing any of this.

Someone not-me with technical chops would be good to have too, of course (pull requests welcome BTW!), but that's currently less of a bottleneck.

**Reply by Evansblue787** [reply]
Thank you so much for the detailed reply! 

Whenever I try to run yarn start I keep getting an error that says Fatal: bad revision "head" and I cant figure out how to get past it

**Reply by abesto** [reply]
Weird, that kinda sounds like you're in an uninitialized git repo with no commits. Are you sure you cloned the repo? (As opposed to, say, downloading the source as a zip)?

**Reply by Evansblue787** [reply]
oh there we go, sorry im kind of learning as i go with some help from my brother!
